ompi/mca/coll/base/coll_base_allreduce.c

```diff
@@ -1376,4 +1376,104 @@ int ompi_coll_base_allreduce_intra_allgather_reduce(const void *sbuf, void *rbuf
     return err;
 }
 
+/*
+ *   ompi_coll_base_allreduce_intra_allgather
+ *
+ *   Function:       use allgather for allreduce operation (for internode)
+ *   Accepts:        Same as MPI_Allreduce()
+ *   Returns:        MPI_SUCCESS or error code
+ *
+ *   Description:    Implements allgather based allreduce aimed to improve internode 
+ *                   allreduce latency: this method takes advantage of the send and 
+ *                   receive can happen at the same time; first step is allgather
+ *                   operation to allow all ranks to obtain the full dataset; the second
+ *                   step is to do reduction on all ranks to get allreduce result. 
+ *
+ *   Limitations:    This method is designed for small message sizes allreduce because it 
+ *                   is not efficient in terms of network bandwidth comparing
+ *                   to gather/reduce/bcast type of approach; it should be combined with
+ *                   allgather bruck's method for larger communicator size.
+ */
+int ompi_coll_base_allreduce_intra_allgather(const void *sbuf, void *rbuf, int count,
+                                           struct ompi_datatype_t *dtype,
+                                           struct ompi_op_t *op,
+                                           struct ompi_communicator_t *comm,
+                                           mca_coll_base_module_t *module)
```

Comment:
This implementation is a generalized case of `ompi_coll_base_allreduce_intra_allgather_reduce` - I think a better approach is to:
- Create a new allgather algorithm where every participant sends and receives from everyone. This is done in the current `ompi_coll_base_allreduce_intra_allgather_reduce`
- Instead of creating `ompi_coll_base_allreduce_intra_allgather`, directly modify `ompi_coll_base_allreduce_intra_allgather_reduce`

This has a clear advantage that:
- We avoid 2 allreduce algorithms doing very similar things
- We have a faster allgather algorithm for long-latency networks, e.g. EFA.

Reply:
The best is to merge the two methods in a separate PR.

Reply:
> The best is to merge the two methods in a separate PR.

No. This is worse.

This PR introduces a new algorithm 8 - do you want to only remove it later? It causes unnecessary churns.

If the concern is complexity, I think you can remove allreduce from this PR and open a new one.

Reply:
ok.  I can add direct messaging as allgather algorithm 9 and remove allreduce algirthm 8

Reply:
I have moved the direct messaging method to allgather as discussed.